Mage.Sets/src/mage/cards/r/RadagastTheBrown.java

```diff
@@ -0,0 +1,128 @@
+
+    static {
+        etbFilter.add(TokenPredicate.FALSE);
+    }
+
+    static final FilterCard cardFilter = new FilterCard("creature card that doesn't share a creature type with a creature you control");
+
+    static {
+        cardFilter.add(CardType.CREATURE.getPredicate());
+        cardFilter.add(SharesACreatureTypeWithCreaturesYouControl.DOES_NOT);
+    }
+
+    public RadagastTheBrown(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{2}{G}{G}");
+        
+        this.supertype.add(SuperType.LEGENDARY);
+        this.subtype.add(SubType.AVATAR);
+        this.subtype.add(SubType.WIZARD);
+        this.power = new MageInt(2);
+        this.toughness = new MageInt(5);
+
+        // Whenever Radagast the Brown or another nontoken creature enters the battlefield under your control,
+        // look at the top X cards of your library, where X is that creature's mana value.
+        // You may reveal a creature card from among them that doesn't share a creature type with a creature you control
+        // and put it into your hand.
+        // Put the rest on the bottom of your library in a random order.
+        this.addAbility(new EntersBattlefieldControlledTriggeredAbility(
+                Zone.BATTLEFIELD,
+                new LookLibraryAndPickControllerEffect(
+                        new TargetManaValue(), 1,
+                        cardFilter,
+                        PutCards.HAND, PutCards.BOTTOM_RANDOM
+                ),
+                etbFilter, false
+        ));
+    }
+
+    private RadagastTheBrown(final RadagastTheBrown card) {
+        super(card);
+    }
+
+    @Override
+    public RadagastTheBrown copy() {
+        return new RadagastTheBrown(this);
+    }
+}
+
+class TargetManaValue implements DynamicValue {
+
+    @Override
+    public int calculate(Game game, Ability source, Effect effect) {
+        if (!(source instanceof EntersBattlefieldAllTriggeredAbility)) {
+            return 0;
+        }
+        UUID enteredId = ((EntersBattlefieldControlledTriggeredAbility) source).getTriggerEvent().getTargetId();
+        Permanent enteredBattlefield = game.getPermanent(enteredId);
+        if (enteredBattlefield == null) {
+            return 0;
+        }
+        return enteredBattlefield.getManaValue();
+    }
+
+    @Override
+    public DynamicValue copy() {
+        return new TargetManaValue();
+    }
+
+    @Override
+    public String getMessage() {
+        return "that creature's mana value";
+    }
+}
+
+enum SharesACreatureTypeWithCreaturesYouControl implements Predicate<Card> {
+    DOES(true),
+    DOES_NOT(false);
+    private final boolean doesShare;
+
+    SharesACreatureTypeWithCreaturesYouControl(boolean doesShare) {
+        this.doesShare = doesShare;
+    }
+    @Override
+    public boolean apply(Card card, Game game) {
+        UUID playerId = card.getOwnerId();
```

Comment:
Should check controller instead of owner?

Reply:
No, this one is owner because it's a `Card`, which doesn't have a controller. Maybe if this predicate should get its own class it might be a more general predicate that works on permanents too - not sure if that's advisable. Might be best to just call it `RadagastPredicate` and be done with it.

Reply:
Oops. You are correct.

Reply:
Actually I did it with an anonymous function